include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,485 @@
+ * \brief Contains a gap decorator to annotate sequences with gaps using a set.
+ */
+
+#pragma once
+
+#include <set>
+#include <type_traits>
+
+#include <range/v3/all.hpp>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/alphabet/gap/gapped.hpp>
+#include <seqan3/range/container/concept.hpp>
+#include <seqan3/range/detail/random_access_iterator.hpp>
+
+// ------------------------------------------------------------------
+// gap_decorator_anchor_set
+// ------------------------------------------------------------------
+
+namespace seqan3 {
+
+/*!\brief A gap decorator allows the annotation of sequences with gap symbols
+ * leaving the underlying sequence unmodified.
+ * \tparam inner_type The type of alphabet sequences that will be stored.
+ * \implements seqan3::aligned_sequence_concept.
+ * \ingroup decorator
+ *
+ * This class may be used whenever you want to store or compute an alignment.
+ * The underlying (ungapped sequence) remains unmodified, and is augmented by
+ * gap information. Iterating over a gap_decorator sequences behaves as if you
+ * had a rolled out, aligned sequence with alphabet and gap symbols.
+ *
+ * \details
+ * The gap_decorator_anchor_set is a modified anchor list approach - instead
+ * of storing tuples of anchor positions and gap lengths relative to the
+ * underlying sequence position, the anchor addresses are virtual and gap
+ * lengths accumulated from left to right, i.e. an anchor gap stores at the 2nd
+ * position all previous gap lengths including its own.
+ * This reduces the lookup time to log (k) with k
+ * being the number of continuous gaps (not gap symbols) and comes at the price
+ * of O(k) worst-case runtime for gap modifications - when inserting or erasing
+ * gaps the tailing gaps have to be updated by the resulting offset.
+ * The gap maintaining structure is a set that is given a gap comparing structure.
+ * Sets are implemented as red-black trees and perform random read operations in
+ * log (k) time. The anchor set approach provides a good trade-off when using
+ * both operator[] and gap insertion/erasure.
+ */
+template <typename inner_type>
+//!\cond
+    requires container_concept<inner_type>
+//!\endcond
+struct gap_decorator_anchor_set
+{
+    //!\privatesection
+    //!\brief The gap type as a tuple storing position and accumulated gap lengths.
+    using gap_t = typename std::pair<size_t, size_t>;
+    //!\brief NULL symbol as placeholder for gap_t initialization.
+    #define _ 0
+
+    /*!\brief Structure allowing the comparison of gaps.
+     * \details It is assumed that the gap structure is always healthy, i.e.
+     * there are no two gaps that are overlapping. The ordering of gaps structures
+     * is exclusively dependent on the gap starting position which is the
+     * ordering criterion for the anchor set implemented as ordered red-black tree.
+    */
+    template <typename gap_t>
+    struct gap_compare {
```

Comment:
The anchor set is initialized with it to explicitly define ordering on the first tuple element (not both).

Reply:
But it does not matter since you never have the case that two values of the first tuple entry are the same (as you do not have two gaps at the same position). So this is basically the same as the default case.

Reply:
OK, I agree, just looked up the definition in the standard - the behavior is that only when there is a difference in the first position, the second position is compared like lexicographical ordering. I was not sure if there is a guarantee for this behavior, so explicitly handed over a comparator struct..